File: eel-core/src/main/scala/io/eels/component/hive/HiveFilePart.scala
package io.eels.component.hive

import io.eels._
import org.apache.hadoop.fs.{FileSystem, LocatedFileStatus}

class HiveFilePart(dialect: HiveDialect,
                   file: LocatedFileStatus,
                   partition: Partition,
                   metastoreSchema: Schema,
                   schema: Schema,
                   partitionKeys: Seq[String])
                  (implicit fs: FileSystem) extends Part {

  override def reader: SourceReader = {

    // the schema we send to the reader must have any partitions removed, because those columns won't exist
    // in the data files. This is because partitions are not written and instead inferred from the hive meta store.
    val dataSchema = Schema(schema.columns.filterNot(partitionKeys contains _.name))
    val reader = dialect.reader(file.getPath, metastoreSchema, dataSchema)

    new SourceReader {
      override def close(): Unit = reader.close()

      // when we read a row back from the dialect reader, we must repopulate any partition columns requested.
      // Again because those values are not stored in hive, but inferred from the meta store
      override def iterator: Iterator[InternalRow] = reader.iterator.map { row =>
        val map = RowUtils.toMap(dataSchema, row)
        schema.columnNames.map { columnName =>
          map.getOrElse(columnName, partition.get(columnName).orNull)
        }
      }
    }
  }
}

File: eel-core/src/main/kotlin/io/eels/component/hive/HiveFilePart.kt
package io.eels.component.hive

import io.eels.Row
import io.eels.component.Part
import io.eels.component.Predicate
import io.eels.schema.Schema
import io.eels.util.Option
import org.apache.hadoop.fs.FileSystem
import org.apache.hadoop.fs.LocatedFileStatus
import rx.Observable

/**
 * @metastoreSchema the schema as present in the metastore and used to match up with the raw data in dialects
 * where the schema is not present. For example with a CSV format in Hive, the metastoreSchema is required
 * in order to know what each column represents. We can't use the projection schema for this because the projection
 * schema might be in a different order.
 *
 * @projectionSchema the schema actually required, optional in which case the metastoreSchema will be used.
 * The reason the projectionSchema is pushed down to the dialects rather than being applied after is because
 * some file schemas can read data more efficiently if they know they can omit some fields (eg Parquet).
 *
 * @predicate is pushed down to the parquet reader for efficiency
 *
 * @partitions a list of partition key-values for this file. We require this to repopulate the partition
 * values when creating the final Row.
 */
class HiveFilePart(val dialect: HiveDialect,
                   val file: LocatedFileStatus,
                   val metastoreSchema: Schema,
                   val projectionSchema: Schema,
                   val predicate: Option<Predicate>,
                   val partitions: List<PartitionPart>,
                   val fs: FileSystem) : Part {

  override fun data(): Observable<Row> {
    require(projectionSchema.fieldNames().all { it == it.toLowerCase() }, { "B Use only lower case field names with hive" })

    val map = partitions.map { it.key to it.value }.toMap()

    // the schema we send to the dialect must have any partitions removed, because those fields won't exist
    // in the data files. This is because partitions are not written and instead inferred from the hive meta store.
    val projectionWithoutPartitions = Schema(projectionSchema.fields.filterNot { map.containsKey(it.name) })
    val reader = dialect.read(file.path, metastoreSchema, projectionWithoutPartitions, predicate, fs)

    // since we removed the partition fields from the target schema, we must repopulate them after the read
    return reader.map { rowRepopulator(it, projectionSchema, map) }
  }
}

/**
 * Accepts a Row and reformats it according to the target schema, using the lookup map for the missing values.
 */
fun rowRepopulator(row: Row, targetSchema: Schema, lookup: Map<String, Any> = emptyMap()): Row {
  val values = targetSchema.fieldNames().map { fieldName ->
    when {
      lookup.containsKey(fieldName) -> lookup[fieldName]
      else -> row.get(fieldName)
    }
  }
  return Row(targetSchema, values)
}